src/Storages/StorageKeeperMap.cpp

```diff
@@ -0,0 +1,329 @@
+#include <Common/ZooKeeper/KeeperException.h>
+#include <Common/ZooKeeper/Types.h>
+#include "Parsers/ASTExpressionList.h"
+#include "Parsers/ASTFunction.h"
+#include <Parsers/ASTIdentifier.h>
+#include <Interpreters/evaluateConstantExpression.h>
+#include "Parsers/ASTSelectQuery.h"
+#include "Storages/MergeTree/IMergeTreeDataPart.h"
+#include <Core/NamesAndTypes.h>
+#include <Storages/ColumnsDescription.h>
+#include <Storages/checkAndGetLiteralArgument.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+    extern const int BAD_ARGUMENTS;
+}
+
+StorageKeeperMap::StorageKeeperMap(std::string_view keeper_path_, ContextPtr context, const StorageID & table_id)
+    : IStorage(table_id), keeper_path(keeper_path_), zookeeper_client(context->getZooKeeper()->startNewSession())
+{
+    StorageInMemoryMetadata storage_metadata;
+    storage_metadata.setColumns(ColumnsDescription{getNamesAndTypes()});
+    setInMemoryMetadata(storage_metadata);
+
+    if (keeper_path.empty())
+        throw Exception("keeper_path should not be empty", ErrorCodes::BAD_ARGUMENTS);
+    if (!keeper_path.starts_with('/'))
+        throw Exception("keeper_path should start with '/'", ErrorCodes::BAD_ARGUMENTS);
+
+    if (keeper_path != "/")
+    {
+        LOG_TRACE(&Poco::Logger::get("StorageKeeperMap"), "Creating root path {}", keeper_path);
+
+        size_t cur_pos = 0;
+        do
+        {
+            size_t search_start = cur_pos + 1;
+            cur_pos = keeper_path.find('/', search_start);
+            if (search_start == cur_pos)
+                throw Exception("keeper_path is invalid, contains subsequent '/'", ErrorCodes::BAD_ARGUMENTS);
+
+            auto path = keeper_path.substr(0, cur_pos);
+            auto status = getClient()->tryCreate(path, "", zkutil::CreateMode::Persistent);
+            if (status != Coordination::Error::ZOK && status != Coordination::Error::ZNODEEXISTS)
+                throw zkutil::KeeperException(status, path);
+        } while (cur_pos != std::string_view::npos);
+    }
+}
+
+NamesAndTypesList StorageKeeperMap::getNamesAndTypes()
+{
+    return {{"key", std::make_shared<DataTypeString>()}, {"value", std::make_shared<DataTypeString>()}};
+}
+
+class StorageKeeperMapSink : public SinkToStorage
+{
+    StorageKeeperMap & storage;
+    std::unordered_map<std::string, std::string> new_values;
+
+public:
+    StorageKeeperMapSink(const Block & header, StorageKeeperMap & storage_) : SinkToStorage(header), storage(storage_) { }
+
+    std::string getName() const override { return "StorageKeeperMapSink"; }
+
+    void consume(Chunk chunk) override
+    {
+        auto block = getHeader().cloneWithColumns(chunk.getColumns());
+
+        size_t rows = block.rows();
+        for (size_t i = 0; i < rows; ++i)
+        {
+            std::string key = block.getByPosition(0).column->getDataAt(i).toString();
+
+            if (key.find('/') != std::string::npos)
```

Comment:
Let's hash keys with SipHash128 and encode with base64, and put all the columns serialized in the data?